src/core/shape_inference/include/segment_max_shape_inference.hpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cmath>
+
+#include "openvino/op/segment_max.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace v16 {
+template <class TShape, class TRShape = result_shape_t<TShape>>
+std::vector<TRShape> shape_infer(const SegmentMax* op,
+                                 const std::vector<TShape>& input_shapes,
+                                 const ITensorAccessor& tensor_accessor = make_tensor_accessor()) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 2 || input_shapes.size() == 3);
+
+    // validate data input
+    const auto& data_shape = input_shapes[0];
+    const auto is_data_shape_rank_static = data_shape.rank().is_static();
+    if (is_data_shape_rank_static) {
+        NODE_SHAPE_INFER_CHECK(op, input_shapes, data_shape.size() > 0, "The data input cannot be a scalar.");
+    }
+
+    // validate segment_ids input
+    const auto& segment_ids_shape = input_shapes[1];
+    const auto is_segment_ids_rank_static = segment_ids_shape.rank().is_static();
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           segment_ids_shape.rank().compatible(1),
+                           "segment_ids must be a 1D input. Got: ",
+                           segment_ids_shape);
+    if (is_segment_ids_rank_static && is_data_shape_rank_static) {
+        NODE_SHAPE_INFER_CHECK(op,
+                               input_shapes,
+                               data_shape[0].compatible(segment_ids_shape[0]),
+                               "The number of elements in segment_ids must match the first dimension of data.");
+    }
+    const auto segment_ids = ov::op::get_input_const_data_as<TRShape, int64_t>(op, 1, tensor_accessor);
+    if (segment_ids) {
+        NODE_VALIDATION_CHECK(op,
+                              std::is_sorted(segment_ids->begin(), segment_ids->end()),
+                              "segment_ids must be sorted.");
+    }
+
+    // validate num_segments input
+    const auto num_segments_available = op->inputs().size() == 3;
+    const auto num_segments = num_segments_available
+                                  ? ov::op::get_input_const_data_as<TRShape, int64_t>(op, 2, tensor_accessor)
```

Comment:
As this value is going to be a dimension, usage of     `get_input_const_data_as_shape` can be considered.

Reply:
Done.